Improve lib template on Travis

bltavares opened the pull request:
This commit contains changes on the generated lib template to build
cross-platform libs. With this changes, I was able to setup a [Travis]
build with GitHub [Releases] of a cross-platform lib.

It is not possible yet to build a different crate type only using
command line args. It requires a modification on `Cargo.toml` to include
new types of library outputs. I've already opened an issue on [cargo](https://github.com/rust-lang/cargo/issues/6160) to see what should be the case here.

Meanwhile, lib authors must change `Cargo.toml` and include the extra
`crate-type` attribute with all the libs. Sadly, this also means that
`LTO` optimisation is not available if the lib uses `lib` or `rlib`
attributes.

To avoid modifying the `Cargo.toml`, the sugestion would be that on a
future PR we could add to the deploy script a modification on the
`Cargo.toml` manifest to include the corresponding crate-type, and only
such crate type for that target. This means we would be able to bring
LTO and output smaller libs.

[Travis]: https://travis-ci.org/bltavares/rust-over-jna-example/builds/439439854
[Releases]: https://github.com/bltavares/rust-over-jna-example/releases/tag/initial

<!--
Thanks for creating a Pull Request 😄 ! Before you submit, please read the following:
- Read our CONTRIBUTING.md file before submitting a patch.
- By making a contribution, you agree to our Developer Certificate of Origin.
-->

**Choose one:** is this a 🐛 bug fix, a 🙋 feature, or a 🔦 documentation change?
🙋

<!-- Provide a general summary of the changes in the title above -->

## Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] tests pass
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added

## Context
- https://github.com/yoshuawuyts/crossgen/issues/11

## Semver Changes
<!-- Which semantic version change would you recommend? -->

bltavares committed:
Improve lib template on Travis

This commit contains changes on the generated lib template to build
cross-platform libs. With this changes, I was able to setup a [Travis]
build with GitHub [Releases] of a cross-platform lib.

It is not possible yet to build a different crate type only using
command line args. It requires a modification on `Cargo.toml` to include
new types of library outputs. I've already opened an issue on [cargo](https://github.com/rust-lang/cargo/issues/6160) to see what should be the case here.

Meanwhile, lib authors must change `Cargo.toml` and include the extra
`crate-type` attribute with all the libs. Sadly, this also means that
`LTO` optimisation is not available if the lib uses `lib` or `rlib`
attributes.

To avoid modifying the `Cargo.toml`, the sugestion would be that on a
future PR we could add to the deploy script a modification on the
`Cargo.toml` manifest to include the corresponding crate-type, and only
such crate type for that target. This means we would be able to bring
LTO and output smaller libs.

Realated to:
- https://github.com/yoshuawuyts/crossgen/issues/11

[Travis]: https://travis-ci.org/bltavares/rust-over-jna-example/builds/439439854
[Releases]: https://github.com/bltavares/rust-over-jna-example/releases/tag/initial

Signed-off-by: Bruno Tavares <connect+github@bltavares.com>

yoshuawuyts commented on templates/lib/before_deploy.sh:
```diff
@@ -6,7 +6,8 @@ set -ex
   local src=$(pwd) \
-    stage=
+        stage= \
+        linking_args=
```
could we get 2 lines indentation here?

bltavares replied:
I've updated emacs to use 2 spaces as identation, but it keeps wanting to align the variables instead of indenting it.
Should they be indented instead of aligned then?

yoshuawuyts replied:
Yeah, I don't align any of my code; it's usually harder to have people collaborate on (unless it's enforced by something like `gofmt` / `rustfmt`).

yoshuawuyts commented on templates/lib/before_deploy.sh:
```diff
@@ -19,8 +20,29 @@ main() {
+      *)
+          cp target/$TARGET/release/lib$PKG_NAME.so $stage/
+          ;;
+  esac
```
could we get 2 lines indentation here?

yoshuawuyts reviewed (commented):
Overall this looks pretty good to me! -- few small nits, but think we should be good to merge after!

bltavares commented:
I've used the [shfmt](https://github.com/mvdan/sh#shfmt) CLI to fix indentation. Let me know if there is something else to change.

bltavares committed:
Define the type of artifact to be produced by cargo.

s command `cargo` subcommand is exclusively intended to be used to help
with [working arround](https://github.com/rust-lang/rust/issues/51009)
how crate types are defined, in order to help with cross-platform
builds.

It is currently [not possible](https://github.com/rust-lang/cargo/issues/6160) to define a
single `crate-type` override on `cargo build`, which causes libs
intended to be used on other languages to compile more than one type of
crate.

This commit adds a dependency on this [new CLI app](https://github.com/bltavares/cargo-crate-type) to be able to workaround the limitations of Cargo.

We are now able to enable the `-C lto` optimization again, as we will
only compile a single type of artifact

Signed-off-by: Bruno Tavares <connect+github@bltavares.com>

bltavares commented:
I've had the time to write a small CLI app to help to produce the correct type of lib, it being either static or dynamic.

You can see the resulting run [here](https://travis-ci.org/bltavares/rust-over-jna-example/builds/456710806) and the artifacts [here](https://github.com/bltavares/rust-over-jna-example/releases/tag/crate-type).

Let me know if you would rather have this commits splits, or if there are any concerns with this extra step and dependency

yoshuawuyts commented:
> Let me know if you would rather have this commits splits, or if there are any concerns with this extra step and dependency

I'm not seeing any extra dependencies being pulled in; this seems all fine. I say we merge this!
